x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/ArrowResponse.java

```diff
@@ -0,0 +1,376 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.arrow;
+
+import org.apache.arrow.memory.ArrowBuf;
+import org.apache.arrow.vector.compression.NoCompressionCodec;
+import org.apache.arrow.vector.ipc.ArrowStreamWriter;
+import org.apache.arrow.vector.ipc.WriteChannel;
+import org.apache.arrow.vector.ipc.message.ArrowFieldNode;
+import org.apache.arrow.vector.ipc.message.ArrowRecordBatch;
+import org.apache.arrow.vector.ipc.message.IpcOption;
+import org.apache.arrow.vector.ipc.message.MessageSerializer;
+import org.apache.arrow.vector.types.Types.MinorType;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.bytes.ReleasableBytesReference;
+import org.elasticsearch.common.io.stream.BytesStream;
+import org.elasticsearch.common.io.stream.RecyclerBytesStreamOutput;
+import org.elasticsearch.common.recycler.Recycler;
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.rest.ChunkedRestResponseBodyPart;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+public class ArrowResponse implements ChunkedRestResponseBodyPart, Releasable {
+
+    public static class Column {
+        private final BlockConverter converter;
+        private final String name;
+
+        public Column(String esqlType, String name) {
+            this.converter = ESQL_CONVERTERS.get(esqlType);
+            if (converter == null) {
+                throw new IllegalArgumentException("ES|QL type [" + esqlType + "] is not supported by the Arrow format");
+            }
+            this.name = name;
+        }
+    }
+
+    private final List<Column> columns;
+    private Iterator<ResponseSegment> segments;
+    private ResponseSegment currentSegment;
+
+    public ArrowResponse(List<Column> columns, List<Page> pages) {
+        this.columns = columns;
+
+        currentSegment = new SchemaResponse(this);
+        List<ResponseSegment> rest = new ArrayList<>(pages.size());
+        for (int p = 0; p < pages.size(); p++) {
+            rest.add(new PageResponse(this, pages.get(p)));
+        }
+        rest.add(new EndResponse(this));
+        segments = rest.iterator();
+    }
+
+    @Override
+    public boolean isPartComplete() {
+        return currentSegment == null;
+    }
+
+    @Override
+    public boolean isLastPart() {
+        // Even if sent in chunks, the entirety of ESQL data is available, so it's single (chunked) part
+        return true;
```

Comment:
In a follow-up it'd be cool to change this. Arrow's IPC seems designed to work properly with streaming these results. We haven't plumbed that out of Elasticsearch though. And it's hard! Certainly not for this PR.

Reply:
Sorry, I meant change it so we don't have to wait for all of the results to arrive. Nothing to do for now, there's lots of plumbing to make it work.